### PR TITLE
(WIP) Enable PGO builds for psmpi

### PR DESCRIPTION
**Pull request by damianam**
This is relevant just for KNL systems

### CODE REVIEW

File: easybuild/easyblocks/p/psmpi.py [hunk 1]
```diff
@@ -29,12 +29,60 @@
...
+/****************************************
+ * Buggy, ugly and extremely simple MPI ping pong.
+ * The goal is to trigger pscom internal routines
+ * to generate profiles to be used by PGO-enabled
+ * compilers.
+ *
+ * Nothing more, nothing less.
+ *
+ * We try small and large messages to walk the path for
+ * eager and rendezvous protocols
```
Comment (boegel) [inline easybuild/easyblocks/p/psmpi.py]: @damianam Where did this come from? Should mention author or upstream source?

File: easybuild/easyblocks/p/psmpi.py [hunk 2]
```diff
@@ -58,9 +106,56 @@ def extra_options(extra_vars=None):
...
+            'mpiexec_cmd': ['srun -n ', "Command to run benchmarks to generate PGO profile. With -n switch", CUSTOM],
         })
         return extra_vars
 
+    def pgo_steps(self):
+        """
+        Set of steps to be performed after the initial installation, if PGO were enabled
+        """
+        # Remove old profiles
+        prof_dir = '%s/profile/' % self.builddir
```
Comment (boegel) [inline easybuild/easyblocks/p/psmpi.py]: `os.path.join(self.builddir, 'profile')`

File: easybuild/easyblocks/p/psmpi.py [hunk 3]
```diff
@@ -58,9 +106,56 @@ def extra_options(extra_vars=None):
...
+        old_profs = os.listdir(prof_dir)
+        for prof in [os.path.join(prof_dir, p) for p in old_profs]:
+            remove_file(prof)
```
Comment (boegel) [inline easybuild/easyblocks/p/psmpi.py]: `remove_dir(prof_dir)` followed by `mkdir(prof_dir)`?

File: easybuild/easyblocks/p/psmpi.py [hunk 4]
```diff
@@ -58,9 +106,56 @@ def extra_options(extra_vars=None):
...
+        new_profs = os.listdir(prof_dir)
+        if new_profs:
+            raise EasyBuildError("The configure PGO profiles were not wiped (%s)" % new_profs)
```
Comment (boegel) [inline easybuild/easyblocks/p/psmpi.py]: How can this ever be triggered?

File: easybuild/easyblocks/p/psmpi.py [hunk 5]
```diff
@@ -92,6 +187,10 @@ def configure_step(self):
         if self.cfg['mpich_opts'] is not None:
             self.cfg.update('configopts', ' --with-mpichconf="%s"' % self.cfg['mpich_opts'])
 
+        # Add PGO related options, if enabled
+        if self.cfg['pgo']:
+            self.cfg.update('configopts', ' --with-profile=gen --with-profdir=%(builddir)s/profile')
```
Comment (boegel) [inline easybuild/easyblocks/p/psmpi.py]: @damianam Avoid having the profdir location in two places, use a class variable that you define early on (in `__init__`)

File: easybuild/easyblocks/p/psmpi.py (hunk 1 above)
Comment (damianam) [reply easybuild/easyblocks/p/psmpi.py]: There is no upstream source or author other than me. It is a very simple test. I based it on a hello world example, but I don't think it should matter too much......

File: easybuild/easyblocks/p/psmpi.py [hunk 6]
```diff
@@ -29,12 +29,73 @@
 """
 
 import easybuild.tools.toolchain as toolchain
```
Comment (boegel) [inline easybuild/easyblocks/p/psmpi.py]: @damianam While you're at it, please move down this import among the other `easybuild` import statements

File: easybuild/easyblocks/p/psmpi.py [hunk 7]
```diff
@@ -58,9 +124,59 @@ def extra_options(extra_vars=None):
             'mpich_opts': [None, "Optional options to configure MPICH", CUSTOM],
             'threaded': [False, "Enable multithreaded build (which is slower)", CUSTOM],
             'pscom_allin_path': [None, "Enable pscom integration by giving its source path", CUSTOM],
+            'pgo': [False, "Enable profiling guided optimizations", CUSTOM],
+            'mpiexec_cmd': ['srun -n ', "Command to run benchmarks to generate PGO profile. With -n switch", CUSTOM],
```
Comment (akesandgren) [inline easybuild/easyblocks/p/psmpi.py]: mpirun is a better default, that usually works even outside the batch system.

File: easybuild/easyblocks/p/psmpi.py [hunk 8]
```diff
@@ -58,9 +124,59 @@ def extra_options(extra_vars=None):
...
+
+        # Compile and run example to generate profile
+        print_msg("generating PGO profile...")
+        (out, _) = run_cmd('%s 2 hostname' % self.cfg['mpiexec_cmd'])
+        nodes = out.split()
+        if nodes[0] == nodes[1]:
+            raise EasyBuildError("The profile is generated with 1 node! Use 2 nodes to generate a proper profile!")
+
+        write_file('pingpong.c', PINGPONG_PGO_TEST)
+        run_cmd('%s/bin/mpicc pingpong.c -o pingpong' % self.installdir)
```
Comment (akesandgren) [inline easybuild/easyblocks/p/psmpi.py]: os.path.join(self.installdir, 'bin', 'mpicc') or perhaps the correct ENV var for MPICC

File: easybuild/easyblocks/p/psmpi.py (hunk 7 above)
Comment (ocaisa) [reply easybuild/easyblocks/p/psmpi.py]: `mpiexec -np` is the only one that is actually part of the standard, if you were to go for a generic default then it should really be that

**Comment by akesandgren** [comment]
@damianam ping

**Comment by damianam** [comment]
I really dropped the ball here, but it is not relevant anymore, so closing.

**Pull request closed without merging**